### PR TITLE
cache: reflector should never stop watching

### PR DESCRIPTION
**Pull request by xiang90**
A recent change tries to separate resync and relist. The motivation
was to avoid triggering relist when a resync is required.

However, the change is not effective since it stops the watcher. As hongchao
mentioned in the original comment, today's storage interface will not deliever
any progress notification to the watch chan. So any watcher that does not receive
events for the last few seconds will not be able to catch up from the previous
index after a hard close since the index of the last received event is out of
the cache window inside etcd2.

This pull request tries to fix this issue by not stoping watcher when a resync is
required.

/cc @hongchaodeng @wojtek-t @timothysc @rrati @smarterclayton

### CODE REVIEW

**Comment by xiang90** [comment]
There is a race on r.timeoutForWatch(). I think if we want to take this patch, then we can actually remove the timeout since resync is not relevant to (will not break) watch anymore.

**Comment by wojtek-t** [comment]
>
There is a race on r.timeoutForWatch(). I think if we want to take this patch, then we can actually remove the timeout since resync is not relevant to (will not break) watch anymore.

This is partially true. We will want to have a timeout on watch to avoid situations (that should never happen, but still), when some hanging/broken watch will make the client to not receive anything forever.

But I completely agree that it no longer has to be that complicated.
Actually, with that change it should be:

    Timeout: time.Duration(float64(minWatchTimeout) (rand.Float64() + 1.0))

Can you please change that too?

File: pkg/client/cache/reflector.go [hunk 1]
```diff
@@ -306,13 +306,34 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
...
 
+	resyncerrc := make(chan error, 1)
+	go func() {
+		for {
+			select {
+			case <-resyncCh:
+			case <-stopCh:
+				return
+			}
+			if r.canForceResyncNow() {
```
Comment (wojtek-t) [inline pkg/client/cache/reflector.go]: This is no longer needed in that case.
If you received sth from resyncCh, that means that you really want to do a resync.

So you should do the resync unconditionally here.

File: pkg/client/cache/reflector.go [hunk 2]
```diff
@@ -306,13 +306,34 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
...
+
 	for {
-		options := api.ListOptions{
+		options = api.ListOptions{
 			ResourceVersion: resourceVersion,
 			// We want to avoid situations when resyncing is breaking the TCP connection
 			// - see comment for 'timeoutForWatch()' for more details.
 			TimeoutSeconds: r.timeoutForWatch(),
 		}
+
```
Comment (wojtek-t) [inline pkg/client/cache/reflector.go]: As I wrote in the PR comment - with this approach timeoutForWatch can be signiifcantly simplied.
So it should be just:

     time.Duration(float64(minWatchTimeout) * (rand.Float64() + 1.0))

and we don't even need a function for it.
It was necessary when relist was strictly connected to resync (i.e. we couldnt' do resync with watch opened).

**Comment by wojtek-t** [comment]
@xiang90 - thanks a lot for catching this. I added some comments. If you can fix those asap it would be great - I would like to have this PR for 1.3

**Comment by rrati** [comment]
Maybe I'm missing something, but how will these changes prevent the watch from being broken?  The timeout on the watch will still cause the watch to break.  I think extending the timeout and decoupling the timeout from the resync timer makes sense, but if the stated goal is to not break the watch then I don't see how this will do that.

Also, these are kube watches and not etcd watches.  Events won't be lost when the watch is broken.  When the watch is re-established all events from watch version->current will be sent.

**Comment by timothysc** [comment]
> However, the change is not effective since it stops the watcher. 

As mentioned in the other issue, the kube watch(through the cacher) behavior can break and reconnect with the old resource version and not miss a beat.

**Comment by wojtek-t** [comment]
@timothysc @rrati - what you wrote above is correct, but with a caveat

To clarify - there is a limitted history (both in cacher than in etcd). E.g. in cacher for pods we have 1000 recent changes. So if between breaking the previous watch and sending the next one there will happen more than 1000 creations/updates/deletions of different pods, then you will get "out of history" error (and you will be forced to relist and start watch from that point). This is exactly the situation which we want to avoid.

If Resync() is long, the the period of time when no watch is opened may potentially be too long. This PR is supposed to address it.

@xiang90 - we need to be aware that Resync() may block the underlying cache, so watch event may not be able to be processed for some time. But I still think what you suggested is better.

**Comment by xiang90** [comment]
@rrati @timothysc @wojtek-t This is relevant since I want to avoid all unnecessary watcher stoping throughout all layers. I know probably there are some other places they we stop watchers. If you have a good idea of where, we would love to fix them all.


There are several reasons for this:

1. braking watchers is blocking processing. Stoping a watcher at any layer will have side-effect to the bottom layer cache and finally cause etcd watcher to drop. 

2. right now, we might drop watchers if they do not receive anything for x minutes. This can be false positive. There are situations that there were just no events in x minutes and the connection were all fine. Soon we can get rid of all hard timeouts on watchers unless there is a real watcher issue down to etcd server. etcd3 will push progress notifications periodically to watchers.

**Comment by timothysc** [comment]
I'm not opposed to the idea, we wanted to not break in the other PR too.

**Comment by xiang90** [comment]
@wojtek-t Yes. I know rsync can block the store thus blocking watch. We should probably give all watchers a careful look and try to make them "unblocking" if possible (explicitly start queuing up the coming events before a known blocking situation).

**Comment by xiang90** [comment]
> I'm not opposed to the idea, we wanted to not break in the other PR too.

We will get there eventually.

**Comment by xiang90** [comment]
@wojtek-t fixed

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #25539

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #25539
[logs](https://00e9e64bac8c5d222ddd2cf4343ef78e8aa74a19e620208053-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F25945%2Fkubernetes-pull-test-unit-integration%2F26970%2Fbuild-log.txt?qk=AD5uMEuI3Ty4RtTgjalLLFyuqrsMf2MwkUWthXTrkPQUT6EE8105vZE0EW87jJkzw0EOYJVR-L8Vlao9TFmUKfNzO75SBIcRq94kwWw6AHccbE6zg-DVaBDvl_i3RUuRNZdsLJpTqm8wofuJcC2wJWmGraNv_sc6b2qMUqV62jbPrFOMo6rx5nffnoiTALUneSIEAMwu7oF9L4EffhvokhHG1_MABhYMsUmZEDbGq7DDhlX1ANoAW-UdURlnmbnxVHlRVcHR3_fsAEfdRNWM9sV4b2WMKr7HDgeH3Q8JLOD2YamDG_0zOPGo1ZswogA4hS7Nq8v-DP43oQQfaH3sYu370tMFw7GEDlRZeHHD22LBBbtS33RYJRR2dW8e6NEIVZqjPmwjj6I922u5Tt6YDsbUCWBXfUAU8xZff6FycfKbh-zT29frd-dhy1OmhQV2evIbYjQWwEaqHKOuMWNk7vBaOQoDAKpgix5CwDP86y1x1VBBvEG-oa4bCKiplZTA3w9tWndgSY8KoCpRb6brUXWWc8RJyySakrv8XkNqd_arMTpWxWA-E9ST2toQKwnHHh9QS4M1JdkBVabcZnXdAg9VGizWpxxPFPFsL2weVMr3ZFz_HE8oSCOCbjQYRA7p5VPmEFjFyuw2UpS3RkN0YUdARih-WMU9rvaywPS9IqGyGByOV1uvWRSuOf6xn4TF1eQ5XMLUo-yXxSvaMzlELM2ydgtFisvvE-YKa24sHqkEEPQ8ysVTI9v5ZEx7A7v4hBznflERXL2kJt9N-_AjHeuYREap7gyjIc3bgXjgVBHNp4rXQb6-lmcVGft8DhIh1mzuHqjVRD7dqTTV3e6FRVS80TcPhRLjE5IMDkA60tFDiTJfTGLNpWYzxiF9e-jkG28YLa9l2dxG6HC-YDDuzwGz4K773jTbm73wAIVKbtl2CSYPbWqLnAk)

**Comment by wojtek-t** [comment]
>
@wojtek-t Yes. I know rsync can block the store thus blocking watch. We should probably give all watchers a careful look and try to make them "unblocking" if possible (explicitly start queuing up the coming events before a known blocking situation).

Sure that's fine. I just wanted to point on it. 

LGTM - thanks!

**Comment by xiang90** [comment]
@k8s-bot unit test this issue: #25953

**Comment by xiang90** [comment]
@k8s-bot e2e test this issue: #24652

**Comment by xiang90** [comment]
@wojtek-t rebased with master to fix a few flakes. re-apply lgtm.

**Pull request merged**